File: skills/developer/scripts/new.go
// Command new scaffolds the folder/file structure for a Go kernel library.
//
// It is the Go counterpart of scripts/new.py, which scaffolds a Python kernel
// library. Instead of a Python package, this produces an idiomatic Go module:
//
//	<lib>/
//	  go.mod
//	  kernel/kernel.go                     # SysCommand, ExitCode, Exit, Process, Kernel
//	  processes/processes.go               # System Processes
//	  implementations/implementations.go   # System Interfaces (facades)
//	  <lib>.go                             # entrypoint: wires maps + Bootstrap()
//	  kernel_test.go
//	  specs/
//	  README.md
//	  .gitignore
//
// Usage:
//
//	go run scripts/new.go <library_name>
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
)

// kernelCode is the kernel package: the Go translation of files/kernel.py.
const kernelCode = `package kernel

import "fmt"

// SysCommand is how code consuming the kernel interacts with it. Each command
// is mapped to a Process in the entrypoint's command/process map.
type SysCommand int

const (
	TEST SysCommand = iota + 1
)

// ExitCode is the status a Process reports back through its Exit.
type ExitCode int

const (
	GOOD  ExitCode = 0
	ERROR ExitCode = 1
)

// Exit is the value every Process returns. Data carries arbitrary keyword-style
// results; the "msg" key is always expected to be present.
type Exit struct {
	Code ExitCode
	Data map[string]any
}

func (e Exit) String() string {
	return fmt.Sprintf("Exit Code is %d and data is %v", e.Code, e.Data)
}

// Process mirrors the Python process signature: it receives the keyword
// arguments passed to Kernel.Cmd plus the system interfaces, and returns an Exit.
type Process func(data map[string]any, systems map[string]any) Exit

// Kernel dispatches System Commands to their System Processes.
type Kernel struct {
	systemInterfaces map[string]any
	cmdMap           map[SysCommand]Process
}

// Bootstrap builds a Kernel from a command/process map and a system-interface map.
func Bootstrap(cmdMap map[SysCommand]Process, systemMap map[string]any) *Kernel {
	return &Kernel{cmdMap: cmdMap, systemInterfaces: systemMap}
}

// Cmd looks up the Process for cmd and invokes it with data and the system interfaces.
func (k *Kernel) Cmd(cmd SysCommand, data map[string]any) Exit {
	return k.cmdMap[cmd](data, k.systemInterfaces)
}
`

// processesCode is the processes package: where System Processes live.
const processesCode = `package processes

import "{{library_name}}/kernel"

// Test is a placeholder System Process. Processes return a kernel.Exit; the
// "msg" key in Data is always required.
func Test(data map[string]any, systems map[string]any) kernel.Exit {
	return kernel.Exit{Code: kernel.GOOD, Data: map[string]any{"msg": "test process ran"}}
}
`

// implementationsCode is the implementations package: where System Interfaces live.
const implementationsCode = `package implementations

// System Interfaces implement the Facade pattern over systems external to the
// kernel library (databases, file systems, cloud services, API endpoints, ...).
// Define them here and wire them into the entrypoint's systemInterfaces map.
`

// entrypointCode is the library entrypoint: the Go analog of the Python __init__.py.
const entrypointCode = `package {{library_name}}

import (
	"{{library_name}}/kernel"
	"{{library_name}}/processes"
)

// systemInterfaces maps aliases to System Interface implementations.
var systemInterfaces = map[string]any{}

// commandProcessMap maps each System Command to the System Process that handles it.
var commandProcessMap = map[kernel.SysCommand]kernel.Process{
	kernel.TEST: processes.Test,
}

// Bootstrap wires the maps and returns a ready-to-use Kernel.
func Bootstrap() *kernel.Kernel {
	return kernel.Bootstrap(commandProcessMap, systemInterfaces)
}
`

// testCode is a smoke test for the generated library.
const testCode = `package {{library_name}}

import (
	"testing"

	"{{library_name}}/kernel"
)

func TestBootstrapCmd(t *testing.T) {
	k := Bootstrap()
	exit := k.Cmd(kernel.TEST, map[string]any{})
	if exit.Code != kernel.GOOD {
		t.Fatalf("expected GOOD, got %v", exit.Code)
	}
	if _, ok := exit.Data["msg"]; !ok {
		t.Fatal("expected a msg in exit data")
	}
}
`

// goModCode declares the module. The module path equals the library name, so
// internal imports resolve as "<lib>/kernel", "<lib>/processes".
const goModCode = `module {{library_name}}

go 1.22
`

const readmeCode = `# {{library_name}}

A Go kernel library scaffolded by scripts/new.go.

- System Commands live in kernel/kernel.go (the SysCommand enum).
- System Processes live in processes/processes.go.
- System Interfaces live in implementations/implementations.go.
- The entrypoint {{library_name}}.go wires them together via Bootstrap().
`

const gitignoreCode = `/bin/
*.exe
*.test
*.out
`

// renderTemplate substitutes every {{ library_name }} placeholder with libraryName.
// Mirrors render_template in new.py.
func renderTemplate(libraryName, tmpl string) string {
	pattern := regexp.MustCompile(`\{\{\s*library_name\s*\}\}`)
	return pattern.ReplaceAllString(tmpl, libraryName)
}

// isDir reports whether path exists and is a directory.
func isDir(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}

// checkFirst records which of the library's subfolders already exist, so we can
// skip re-creating them. Mirrors check_first in new.py.
func checkFirst(impPath, procPath, specsPath string) map[string]bool {
	return map[string]bool{
		"imp_path":   isDir(impPath),
		"proc_path":  isDir(procPath),
		"specs_path": isDir(specsPath),
	}
}

// writeFile creates parent dirs as needed and writes content, but never clobbers
// an existing file.
func writeFile(path, content string) error {
	if _, err := os.Stat(path); err == nil {
		return nil // already exists; leave it untouched
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	return os.WriteFile(path, []byte(content), 0o644)
}

// makeFolders creates the processes, implementations and specs folders, writing
// each package's source file when the folder was absent. Mirrors make_folders in new.py.
func makeFolders(existing map[string]bool, libraryName, impPath, procPath, specsPath string) error {
	if !existing["imp_path"] {
		if err := os.MkdirAll(impPath, 0o755); err != nil {
			return err
		}
		if err := writeFile(filepath.Join(impPath, "implementations.go"), implementationsCode); err != nil {
			return err
		}
	}

	if !existing["proc_path"] {
		if err := os.MkdirAll(procPath, 0o755); err != nil {
			return err
		}
		if err := writeFile(filepath.Join(procPath, "processes.go"), renderTemplate(libraryName, processesCode)); err != nil {
			return err
		}
	}

	if !existing["specs_path"] {
		if err := os.MkdirAll(specsPath, 0o755); err != nil {
			return err
		}
	}

	return nil
}

// makeKernelInitFiles writes the module-level files: the kernel package, the
// entrypoint, the test, go.mod, README and .gitignore. Mirrors
// make_kernel_init_files in new.py.
func makeKernelInitFiles(libraryName string) error {
	files := map[string]string{
		filepath.Join(libraryName, "kernel", "kernel.go"): kernelCode,
		filepath.Join(libraryName, libraryName+".go"):     renderTemplate(libraryName, entrypointCode),
		filepath.Join(libraryName, "kernel_test.go"):      renderTemplate(libraryName, testCode),
		filepath.Join(libraryName, "go.mod"):              renderTemplate(libraryName, goModCode),
		filepath.Join(libraryName, "README.md"):           renderTemplate(libraryName, readmeCode),
		filepath.Join(libraryName, ".gitignore"):          gitignoreCode,
	}
	for path, content := range files {
		if err := writeFile(path, content); err != nil {
			return err
		}
	}
	return nil
}

// run parses the library name from the command line and scaffolds the library.
// Mirrors run in new.py.
func run() error {
	libraryName := "mykernel"
	if len(os.Args) > 1 && os.Args[1] != "" {
		libraryName = os.Args[1]
	}

	impPath := filepath.Join(libraryName, "implementations")
	procPath := filepath.Join(libraryName, "processes")
	specsPath := filepath.Join(libraryName, "specs")

	existing := checkFirst(impPath, procPath, specsPath)
	if err := makeFolders(existing, libraryName, impPath, procPath, specsPath); err != nil {
		return err
	}
	if err := makeKernelInitFiles(libraryName); err != nil {
		return err
	}

	fmt.Printf("Scaffolded kernel library %q\n", libraryName)
	return nil
}

func main() {
	if err := run(); err != nil {
		fmt.Fprintln(os.Stderr, "error:", err)
		os.Exit(1)
	}
}

File: skills/developer/scripts/new.ts
#!/usr/bin/env -S npx tsx
/**
 * Scaffolds the folder/file structure for a TypeScript kernel library.
 *
 * Counterpart to scripts/new.py (Python) and scripts/new.go (Go). Instead of a
 * Python package or a Go module, this produces an idiomatic TypeScript package:
 *
 *   <lib>/
 *     package.json
 *     tsconfig.json
 *     src/
 *       kernel.ts          # SysCommand, ExitCode, Exit, Process, Kernel
 *       processes.ts       # System Processes
 *       implementations.ts # System Interfaces (facades)
 *       index.ts           # entrypoint: wires maps + bootstrap()
 *     tests/
 *       kernel.test.ts
 *     specs/
 *     README.md
 *     .gitignore
 *
 * Usage:
 *
 *   npx tsx scripts/new.ts <library_name>
 *   # or, on Node >= 23.6 which runs TypeScript natively:
 *   node scripts/new.ts <library_name>
 */
import { existsSync, mkdirSync, statSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";

// kernelCode is the kernel module: the TypeScript translation of files/kernel.py.
const kernelCode = `export enum SysCommand {
  TEST = 1,
}

export enum ExitCode {
  GOOD = 0,
  ERROR = 1,
}

// Exit is the value every Process returns. data carries arbitrary keyword-style
// results; the "msg" key is always expected to be present.
export class Exit {
  code: ExitCode;
  data: Record<string, unknown>;

  constructor(code: ExitCode, data: Record<string, unknown> = {}) {
    this.code = code;
    this.data = data;
  }

  toString(): string {
    return "Exit Code is " + this.code + " and data is " + JSON.stringify(this.data);
  }
}

// Process mirrors the Python process signature: it receives the arguments passed
// to Kernel.cmd plus the system interfaces, and returns an Exit.
export type Process = (
  data: Record<string, unknown>,
  systems: Record<string, unknown>,
) => Exit;

// Kernel dispatches System Commands to their System Processes.
export class Kernel {
  private systemInterfaces: Record<string, unknown>;
  private cmdMap: Map<SysCommand, Process>;

  private constructor(
    cmdMap: Map<SysCommand, Process>,
    systemMap: Record<string, unknown>,
  ) {
    this.cmdMap = cmdMap;
    this.systemInterfaces = systemMap;
  }

  // bootstrap builds a Kernel from a command/process map and a system map.
  static bootstrap(
    cmdMap: Map<SysCommand, Process>,
    systemMap: Record<string, unknown>,
  ): Kernel {
    return new Kernel(cmdMap, systemMap);
  }

  // cmd looks up the Process for cmd and invokes it with data and the interfaces.
  cmd(cmd: SysCommand, data: Record<string, unknown> = {}): Exit {
    const fn = this.cmdMap.get(cmd);
    if (fn === undefined) {
      throw new Error("No process registered for command " + SysCommand[cmd]);
    }
    return fn(data, this.systemInterfaces);
  }
}
`;

// processesCode is the processes module: where System Processes live.
const processesCode = `import { Exit, ExitCode } from "./kernel";

// test is a placeholder System Process. Processes return an Exit; the "msg" key
// in the data is always required.
export function test(
  data: Record<string, unknown>,
  systems: Record<string, unknown>,
): Exit {
  return new Exit(ExitCode.GOOD, { msg: "test process ran" });
}
`;

// implementationsCode is the implementations module: where System Interfaces live.
const implementationsCode = `// System Interfaces implement the Facade pattern over systems external to the
// kernel library (databases, file systems, cloud services, API endpoints, ...).
// Define and export them here, then wire them into index.ts's systemInterfaces map.
export {};
`;

// indexCode is the library entrypoint: the TypeScript analog of the Python __init__.py.
const indexCode = `import { Kernel, Process, SysCommand } from "./kernel";
import { test } from "./processes";

export * from "./kernel";

// systemInterfaces maps aliases to System Interface implementations.
const systemInterfaces: Record<string, unknown> = {};

// commandProcessMap maps each System Command to the System Process that handles it.
const commandProcessMap: Map<SysCommand, Process> = new Map([
  [SysCommand.TEST, test],
]);

// bootstrap wires the maps and returns a ready-to-use Kernel.
export function bootstrap(): Kernel {
  return Kernel.bootstrap(commandProcessMap, systemInterfaces);
}
`;

// testCode is a smoke test for the generated library.
const testCode = `import { test as nodeTest } from "node:test";
import assert from "node:assert/strict";

import { bootstrap } from "../src/index";
import { ExitCode, SysCommand } from "../src/kernel";

nodeTest("bootstrap runs the TEST command", () => {
  const kernel = bootstrap();
  const exit = kernel.cmd(SysCommand.TEST, {});
  assert.equal(exit.code, ExitCode.GOOD);
  assert.ok("msg" in exit.data);
});
`;

const packageJsonCode = `{
  "name": "{{library_name}}",
  "version": "0.1.0",
  "description": "A TypeScript kernel library.",
  "type": "module",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "build": "tsc -p tsconfig.json",
    "typecheck": "tsc -p tsconfig.test.json",
    "test": "node --import tsx --test tests/**/*.test.ts"
  },
  "devDependencies": {
    "@types/node": "^20.11.0",
    "tsx": "^4.7.0",
    "typescript": "^5.4.0"
  }
}
`;

// tsconfigCode is the build config: emits only src/ to dist/ so that the
// package.json "main" (dist/index.js) resolves correctly.
const tsconfigCode = `{
  "compilerOptions": {
    "target": "ES2022",
    "module": "ESNext",
    "moduleResolution": "Bundler",
    "strict": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "declaration": true,
    "outDir": "dist",
    "rootDir": "src",
    "types": ["node"]
  },
  "include": ["src/**/*.ts"]
}
`;

// tsconfigTestCode type-checks src/ and tests/ together without emitting.
const tsconfigTestCode = `{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "noEmit": true,
    "rootDir": "."
  },
  "include": ["src/**/*.ts", "tests/**/*.ts"]
}
`;

const readmeCode = `# {{library_name}}

A TypeScript kernel library scaffolded by scripts/new.ts.

- System Commands live in src/kernel.ts (the SysCommand enum).
- System Processes live in src/processes.ts.
- System Interfaces live in src/implementations.ts.
- The entrypoint src/index.ts wires them together via bootstrap().

## Usage

    npm install
    npm run typecheck
    npm test
`;

const gitignoreCode = `node_modules/
dist/
*.tsbuildinfo
`;

// renderTemplate substitutes every {{ library_name }} placeholder with libraryName.
// Mirrors render_template in new.py.
function renderTemplate(libraryName: string, tmpl: string): string {
  return tmpl.replace(/\{\{\s*library_name\s*\}\}/g, libraryName);
}

// isDir reports whether path exists and is a directory.
function isDir(path: string): boolean {
  return existsSync(path) && statSync(path).isDirectory();
}

// checkFirst records which of the library's subfolders already exist, so we can
// skip re-creating them. Mirrors check_first in new.py.
function checkFirst(
  srcPath: string,
  testsPath: string,
  specsPath: string,
): Record<string, boolean> {
  return {
    src_path: isDir(srcPath),
    tests_path: isDir(testsPath),
    specs_path: isDir(specsPath),
  };
}

// writeFile creates parent dirs as needed and writes content, but never clobbers
// an existing file.
function writeFile(path: string, content: string): void {
  if (existsSync(path)) {
    return; // already exists; leave it untouched
  }
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, content);
}

// makeFolders creates the src, tests and specs folders, writing the process,
// implementation and test files when their folder was absent. Mirrors
// make_folders in new.py.
function makeFolders(
  existing: Record<string, boolean>,
  libraryName: string,
  srcPath: string,
  testsPath: string,
  specsPath: string,
): void {
  if (!existing.src_path) {
    mkdirSync(srcPath, { recursive: true });
    writeFile(join(srcPath, "processes.ts"), processesCode);
    writeFile(join(srcPath, "implementations.ts"), implementationsCode);
  }

  if (!existing.tests_path) {
    mkdirSync(testsPath, { recursive: true });
    writeFile(join(testsPath, "kernel.test.ts"), testCode);
  }

  if (!existing.specs_path) {
    mkdirSync(specsPath, { recursive: true });
  }
}

// makeKernelInitFiles writes the package-level files: the kernel module, the
// entrypoint, package.json, tsconfig, README and .gitignore. Mirrors
// make_kernel_init_files in new.py.
function makeKernelInitFiles(libraryName: string, srcPath: string): void {
  const files: Record<string, string> = {
    [join(srcPath, "kernel.ts")]: kernelCode,
    [join(srcPath, "index.ts")]: indexCode,
    [join(libraryName, "package.json")]: renderTemplate(libraryName, packageJsonCode),
    [join(libraryName, "tsconfig.json")]: tsconfigCode,
    [join(libraryName, "tsconfig.test.json")]: tsconfigTestCode,
    [join(libraryName, "README.md")]: renderTemplate(libraryName, readmeCode),
    [join(libraryName, ".gitignore")]: gitignoreCode,
  };
  for (const [path, content] of Object.entries(files)) {
    writeFile(path, content);
  }
}

// run reads the library name from the command line and scaffolds the library.
// Mirrors run in new.py.
function run(): void {
  const libraryName = process.argv[2] && process.argv[2] !== "" ? process.argv[2] : "mykernel";

  const srcPath = join(libraryName, "src");
  const testsPath = join(libraryName, "tests");
  const specsPath = join(libraryName, "specs");

  const existing = checkFirst(srcPath, testsPath, specsPath);
  makeFolders(existing, libraryName, srcPath, testsPath, specsPath);
  makeKernelInitFiles(libraryName, srcPath);

  console.log(`Scaffolded kernel library "${libraryName}"`);
}

run();
